### DISCUSSION TITLE
WindowsDriver.cs sporadically triggering 'System.IndexOutOfRange' errors

### DISCUSSION THREAD

**Discussion by dkim-epic** [tui-cs/Terminal.Gui#3033] [Q&A] [answered]
When trying to run a worker (framework copied from the BackgroundWorker UICatalog example) my application will seem to sporadically cause a 'System.IndexOutOfRangeException' that, while crashing the current running application, will sometimes go away if I just restart and rerun my code a few times.

Attached is a screenshot of the error's call stack and my code/class that I'm launching the Dialog box that seems to be triggering the error. I struggled to paste the code in a readable format so I attached it here.

[BackgroundWorkerCode.txt](https://github.com/gui-cs/Terminal.Gui/files/13569233/BackgroundWorkerCode.txt)

**Error:**

![image](https://github.com/gui-cs/Terminal.Gui/assets/78838696/ac3c60e1-1373-478d-90ae-be3a2a6229e8)

While the error does show from the OutputBuffer array, I've also seen it from the contents array:
![image](https://github.com/gui-cs/Terminal.Gui/assets/78838696/2b237a1d-d4e7-4921-acba-61be6d7f4d21)
(From WindowsDriver.cs)

**Comment by tznind** [comment, marked as answer]
When you access UI code from a thread make sure to use `Application.MainLoop.Invoke`

For example

```csharp
Application.Init ();


Task.Run (() => {
	// Runs on another Thread
	Task.Delay (5000).Wait ();

	// Invoke the action onto main UI thread
	Application.MainLoop.Invoke (() => {

		// Your UI code here
		MessageBox.Query ("Heya", "Hey", "ok");
		
	});
});

Application.Run (new Window("yar"));
Application.Shutdown ();
```

**Reply by dkim-epic** [reply]
Thank you!
I added the MainLoop.Invoke wrapper around any logging my child programs do and then after removing the Application.Refresh() method, I've stopped seeing the error!
*Without removing Application.Refresh() I would still get IndexOutOfBound errors.

**Reply by BDisp** [reply]
Because you was forcing the app refreshing on a non UI thread. Thus calling `MainLoop.Invoke` will handle the refresh itself. Only to remind that sometimes you need to call `SetNeedsDisplay` method in the case where the action doesn't call it.

**Reply by dkim-epic** [reply]
Thanks for the extra context, that definitely helps explain why I was triggering the error.
I appreciate all of the help!
